lib/components/admin-command.js

```diff
@@ -0,0 +1,157 @@
+"use strict";
+
+var minimist = require("minimist");
+
+/**
+ * Construct a new AdminCommand helper instance.
+ * @constructor
+ * @param {string} opts.desc A short description to display in the 'help'
+ * command.
+ * @param {AdminCommand~opts=} opts.opts An optional object giving commandline
+ * options recognised by the command.
+ * @param {AdminCommand~func} opts.func The handler function that implements
+ * the command.
+ */
+function AdminCommand(opts) {
+    this.desc = opts.desc;
+    this._func = opts.func;
+
+    if (opts.opts) {
+        this.optspec = {};
+        this.optaliases = {};
+
+        Object.keys(opts.opts).forEach((k) => {
+            var names = k.split(/\|/);
+
+            var result = names.shift().match(/^(!?)(.*?)$/);
+            var name = result[2];
+            var required = result[1];
+
+            this.optspec[name] = {
+                desc: opts.opts[k],
+                required: required,
+            }
+
+            names.forEach((a) => this.optaliases[a] = name);
+        });
+    }
+}
+
+/**
+ * Invoke the handler function for the command. The given list of strings will
+ * be parsed for optional arguments as per the command's opts specification, if
+ * present.
+ * @param {Object} bridge The top-level bridge instance to pass to the handler
+ * function.
+ * @param {string[]} args A list of string arguments parsed from the command
+ * line.
+ * @param {function} respond The responder function to pass to the handler
+ * function.
+ * @return {Promise} Returns the void promise returned by the handler function.
+ */
+AdminCommand.prototype.run = function(bridge, args, respond) {
+    var opts;
+    if (this.optspec) {
+        opts = minimist(args);
+
+        // Canonicalise aliases
+        Object.keys(this.optaliases).forEach((a) => {
+            if (a in opts) {
+                opts[this.optaliases[a]] = opts[a];
+                delete opts[a];
+            }
+        });
+
+        Object.keys(opts).forEach((n) => {
+            if (n === "_") return;
+
+            if (!(n in this.optspec)) {
+                throw Error("Unrecognised option: " + n);
+            }
+        });
+
+        var missing = [];
+        Object.keys(this.optspec).sort().forEach((n) => {
+            if (this.optspec[n].required && !(n in opts)) missing.push(n);
+        });
+
+        if (missing.length) {
+            throw Error("Missing required options: " + missing.join(", "));
+        }
+
+        args = opts._;
+        delete opts["_"];
+    }
+
+    return this._func(bridge, opts, args, respond);
```

Comment:
Right, now I realise what felt wrong about the API shape. We're passing functions around, which in and of itself is fine, but we **never do any async operations**. Why is `respond` A Thing? At a high level, I can see the API shape being:

```js
constructor(opts) {
  // determine your valid flags and descriptions etc
}

process(rawString) {
 // parse string into arguments
 // match them against the command schema in the constructor
 // return them to the caller
}
```

This would make the usage pretty clear too:

```js
var ac = new AdminCommand({
  "verbose": {
    required: false,
    description: "Enable verbose mode",
    aliases: ["v"]
  },
  "url": {
    required: true,
    description: "HTTP URL to hit",
    aliases: ["u"]
  }
});

var cmd = ac.process("--url https://matrix.org -v");
if (cmd.verbose) {
  // turn on verbose mode
}
console.log("URL is %s", cmd.url);
```

This also feels like it should've been done before though..